### DISCUSSION TITLE
Is it possible to open `BuildlessAutoDisposeAsyncNotifier`/`BuildlessAsyncNotifier` for public consumption outside of 'riverpod_generator'?

### DISCUSSION THREAD

**Discussion by abigotado** [rrousselGit/riverpod#1960] [General]
### In continuation of the discussed in https://github.com/rrousselGit/riverpod/discussions/1953

Hello!

I have made a mixin on `AutoDisposeAsyncNotifier` to reuse code in my `AsyncNotifier`s with list pagination logics.

```dart
/// Mixin for pagination screens.
mixin PaginationMixin<T>
    on AutoDisposeAsyncNotifier<PaginationState<T>> {
  /// Fetch list and manage states.
  Future<PaginationState<T>> fetchList({
    final bool isRefresh = false,
    final bool isSocketUpdate = false,
    final String? search,
    final EvaluationStatus? filter,
  }) async {
    if (isSocketUpdate || isRefresh || !(state.value?.isEnd ?? false)) {
      PaginationState<T> pagination = state.value != null
          ? state.value!.copyWith(
              page: isRefresh ? null : state.value!.page,
              items: isRefresh ? null : state.value!.items,
              search: search ?? state.value!.search,
              filter: filter ?? state.value!.filter,
            )
          : PaginationState<T>();

      if (!isSocketUpdate) state = AsyncLoading<PaginationState<T>>();

      debugPrint('isSocketUpdate $isSocketUpdate');

      final int nextPage = (pagination.page ?? 0) + 1;
      state = await AsyncValue.guard(() async {
        final List<T>? list = await getList(
          isSocketUpdate ? 1 : nextPage,
          length: isSocketUpdate ? (state.value?.page ?? 1) * 20 : null,
          isRefresh: isRefresh,
          search: state.value?.search ?? '',
          filter: state.value?.filter ?? EvaluationStatus.all,
        );
        if (list != null) {
          pagination = pagination.copyWith(
            isEnd: list.length < 20,
            items: <T>[
              if (!isSocketUpdate) ...pagination.items ?? <T>[],
              ...list,
            ],
            page: isSocketUpdate ? pagination.page : nextPage,
            search: search ?? pagination.search,
            filter: filter ?? pagination.filter,
          );
        }
        return pagination;
      });
    }
    return state.value ?? PaginationState<T>();
  }

  /// Load list from repository.
  Future<List<T>?> getList(
    final int page, {
    required final EvaluationStatus filter,
    final int? length,
    final bool isRefresh = false,
    final String? search,
  });
}
```

It works well, when I am using `AsyncNotifier` with no additional arguments. But if there are arguments (similar to `family`), generator creates `BuildlessAutoDisposeAsyncNotifier`, not general `AutoDisposeAsyncNotifier`. So the mixin has class mismatch.

This one works well:
```dart
/// The controller of EventsScreen states.
@riverpod
class EventsController extends _$EventsController
    with PaginationMixin<EventModel> {
  @override
  FutureOr<PaginationState<EventModel>> build() {
    return fetchList();
  }

  @override
  Future<List<EventModel>?> getList(
    final int page, {
    required final EvaluationStatus filter,
    final int? length,
    final bool isRefresh = false,
    final String? search,
  }) {
    return ref.read(eventsRepositoryProvider).fetchList(
          page,
          length: length,
          isRefresh: isRefresh,
          search: search,
          filter: filter,
        );
  }
}
```

And this one won't work because of `id` argument:
```dart
/// The controller of [ParticipantsScreen] states.
@riverpod
class ParticipantsController extends _$ParticipantsController
    with PaginationMixin<AccessModel> {
  @override
  FutureOr<PaginationState<AccessModel>> build(final int id) {
    return fetchList();
  }

  @override
  Future<List<AccessModel>?> getList(
    final int page, {
    required final EvaluationStatus filter,
    final int? length,
    final bool isRefresh = false,
    final String? search,
  }) {
    // debugPrint('event controller id: $id');
    return ref.read(participantsRepositoryProvider(id)).fetchList(
          page,
          isRefresh: isRefresh,
          search: search,
          filter: filter,
          length: length,
        );
  }
}
```

The problem would be solved, if there's a possibility to make mixin on `BuildlessAutoDisposeAsyncNotifier` (which is only for internal use now), rather than `AutoDisposeAsyncNotifier`.

So, is it possible to open `BuildlessAutoDisposeAsyncNotifier` for public consumption outside of 'riverpod_generator'? Or maybe here can be another solution?

**Comment by rrousselGit** [comment]
That's not reasonable. It isn't meant to be used

You don't need the `on` clause on mixins for them to work. You can instead do:

```dart
mixin PaginationMixin<T> {
  State<AsyncValue<T>> get state;
  set state(State<AsyncValue<T> value);
  Ref<AsyncValue<T>> get ref;
}
```

If you end up needing this often, you can make an interface

```dart
abstract class AsyncNotifierInterface<T> {
  State<AsyncValue<T>> get state;
  set state(State<AsyncValue<T> value);
  Ref<AsyncValue<T>> get ref;
}

mixin MyMixin<T> implements AsyncNotifierInterface<T> {

}
```

**Comment by rrousselGit** [comment]
I'll maybe expose it in the future, but I'm a bit skeptical about it for now.

It's not something that was designed to be exposed. I'd need to carefully review it before exposing it.  
Considering there's a decent workaround, it's not too important to me.

**Comment by abigotado** [comment]
@rrousselGit Well, I've just thought that it could be possible to expose both Buildless and general Notifiers, if one had already been exposed, but I understand your skepsis about it. Anyway your workaround functions excellent, thank you!

And I just want to specify two more things:

1) Concerning the the `Ref` in a workaround

 If I try to use it like that:
```dart
mixin PaginationMixin<T> {
  Ref<PaginationState<T>> get ref;
```
There is and error, if I try to use the mixin: `'BuildlessAutoDisposeAsyncNotifier.ref' ('AutoDisposeAsyncNotifierProviderRef<PaginationState<AccessModel>> Function()') isn't a valid concrete implementation of 'PaginationMixin.ref' ('Ref<PaginationState<AccessModel>> Function()')`

It is solved, if I create mixin this way:
```dart
mixin PaginationMixin<T> {
  AutoDisposeAsyncNotifierProviderRef<PaginationState<T>> get ref;
```

So it is necessary to specify type of ref, isn't it?

2) Concerning passing a provider to a class as an argument:

```dart
/// Screen with scrollable widget.
class ScrollableScreen<T> extends HookConsumerWidget {
  /// Creates a [ScrollableScreen].
  const ScrollableScreen({
    required this.provider,
    super.key,
  });

  /// Provider of a controller with [PaginationMixin] for the [ScrollableScreen].
  final AutoDisposeAsyncNotifierProvider<dynamic, PaginationState<T>> provider;
```

Is it a proper way? If I change `dynamic` in generic with `AutoDisposeAsyncNotifier<PaginationState<T>>` it will work only with `AutoDisposeAsyncNotifier`, but won't work with  `BuildlessAutoDisposeAsyncNotifier`.

**Comment by rrousselGit** [comment]
> So it is necessary to specify type of ref, isn't it?

Depends on what your mixin needs.

Generally `AutoDisposeRef<T>` vs `Ref<T>` is enough

**Comment by rrousselGit** [comment]
> Concerning passing a provider to a class as an argument:


Providers shouldn't be passed as parameter at all. So I would argue that's asking the wrong question.

**Reply by TimWhiting** [reply]
Honestly, I think I understand why you say this, but I respectfully disagree, though I would recommend passing a `ProviderListenable<T>` in most cases. I'm just trying to understand why you make such a big deal about this. Often I create reusable mutli-selection / selection components by passing a `StateProvider`, or using a combination of a `ProviderListenable` and a callback. The advantage is that I can listen to the provider in the selection component to **only rebuild** that portion of the UI rather than having to wrap lots of things in `Consumer` which just creates more nesting, and I can reuse the selection component many places in the app. I think `ProviderListenable` and a callback should be the recommended way to create such components. The callback doesn't need to listen to the `Provider` but the component does, and it is best to scope the rebuild to the component that actually needs it. Obviously I don't recommend making wrappers for all of flutter's components, but just for the components that your app re-uses in many places with your own custom design style or configuration.

**Reply by abigotado** [reply]
Can it cause some unexpected behavior?

The case is that I have several `AsyncNotifier` providers with lists of different data and pagination. The screens are very similar - they displays a ListView/CustomScrollView with quite the same logic. So I create a basic reusable screen and passing provider (of the specific type) is very convenient and gives me opportunity to reuse more code. Otherwise I have to pass: fetch method, refresh method, loading state, error state, method for listening to error (to show an `AlertDialog`)  - to every screen. The same methods to every screen. And if I pass a provider - I can handle all these methods at the reusable screen.
Why is it bad to do like this?

```dart
/// Screen with scrollable widget.
class ScrollableScreen<T> extends HookConsumerWidget {
  /// Creates a [ScrollableScreen].
  const ScrollableScreen({
    required this.provider,
    required this.itemBuilder,
    this.onErrorRefresh,
    super.key,
  });

  /// Provider of [PaginationMixin] for the [ScrollableScreen].
  final AutoDisposeAsyncNotifierProvider<dynamic, PaginationState<T>> provider;

  /// Called to build children for the list of [T].
  final Widget Function(T) itemBuilder;

  /// Callback for the refresh button of [ErrorBottomSheet].
  final void Function()? onErrorRefresh;

  @override
  Widget build(final BuildContext context, final WidgetRef ref) {
    final AsyncValue<PaginationState<T>> state = ref.watch(provider);

    final PaginationMixin<T> notifier =
        ref.read(provider.notifier) as PaginationMixin<T>;

    final bool isLoading = state.isLoading && state.value?.items != null;
    final bool isError = state.error != null && state.value?.items != null;
    final bool isShimmerShown =
        (state.isLoading && (state.value?.items?.isEmpty ?? true)) ||
            (state.error != null && (state.value?.items?.isEmpty ?? true));
    final int childCount =
        isShimmerShown ? 10 : (state.value?.items?.length ?? 0);

    ref.listen<AsyncValue<PaginationState<T>>>(provider, (
      final AsyncValue<PaginationState<T>>? previousState,
      final AsyncValue<PaginationState<T>> state,
    ) {
      if (state.error != null && state.value?.items == null) {
        showModalBottomSheet<ErrorBottomSheet>(
          isScrollControlled: true,
          context: context,
          builder: (final BuildContext context) {
            debugPrint('ErrorBottomSheet is shown');
            return ErrorBottomSheet(
              errorText: state.error?.toErrorString() ?? '',
              onPressed: () {
                notifier.fetchList(isRefresh: true);
                onErrorRefresh?.call();
              },
            );
          },
        );
      }
    });

    return SomeBeautifulScrollViewWithPagination(itemBuilder, childCount);
    }
    ```

**Reply by rrousselGit** [reply]
@TimWhiting It's like if you tried to return a widget in a Provider. Sure, that'll "work", but it's not exactly recommended either.  
Providers are designed to be used statically. Using them non-statically is not necessarily going to break your app, but that's not how it's meant to be.

Passing `ProviderListenable` is relatively alright. But passing more specific types isn't IMO. The provider types are meant to disappear, and they are meant to become more and more specific. In the future, you're likely to have one "type" of provider per provider (with how `family` works in the code-generator, and how mutations will work when they land).  That'd be hellish if we passed providers around.

Passing them around also hurts devtools and code navigation.


In the end, you could  replace:

```dart
Widget(provider: myProvider)
```

with:

```dart
Widget(state: (ref) => ref.watch(myProvider))
```

**Reply by abigotado** [reply]
@rrousselGit I see your point. Thank you for so clear and wide answer! 

> ```dart
> Widget(state: (ref) => ref.watch(myProvider))
> ```
And that helps very much too.

Also it's possible to pass a notifier the same way:
```dart
notifier: (ref) => ref.read(myProvider.notifier),
```

**Reply by TimWhiting** [reply]
I think Remi's method, and also the `ProviderListenable` + callback method should be mentioned somewhere in the docs as the recommended solution for creating components that can be reusable with multiple different providers for your app. In general I think it is good to know about `ProviderListenable` and the fact that it is the interface that you should be using when creating reusable abstractions on top of Riverpod. I definitely learned pretty quickly that it was not great to pass providers around because then a component would work with one type of provider and not another, for example I couldn't switch a `StateProvider` to a `StateNotifierProvider` in one case.

**Reply by rrousselGit** [reply]
I'll add this to the list.

**Reply by ghost** [reply]
I've recently started using Notifier/AsyncNotifier and after reading some issues and your responses, I kinda get why you've done this, but I think the new approach is too restrictive. StateNotfier is very flexible, it's easy to abstract, compose and partially scope.
Notifier on the other hand is kinda hard to work with if you want to write reusable code or abstract stuff.

Let's say I have a base class for posts feed screen notifier. I might have the same logic on five different screens with slightly different methods. For example all of these posts screens can filter posts, using mixins we can write a PostsFilterMotifier and write a widget which uses this as a scoped provider without passing multiple arguments down the widget tree. With StateNotifier we could write this, for example:
```dart

mixin PostsFilter {
  bool get includeRead;
}

mixin PostsFilterNotifier<S extends PostsFilter> on StateNotifier<S> {
  void changeFilters();
}

class PostsState with PostsFilter {
  PostsState(this.posts, this.includeRead);

  // assume its a full fledged class
  final List<String> posts;
  @override
  final bool includeRead;

  static PostsState initial = PostsState([], true);
}

class PostsNotifier extends StateNotifier<PostsState> with PostsFilterNotifier {
  PostsNotifier(super.state);

  Future<void> loadPosts() async {}

  @override
  void changeFilters() {}
}

// there might be several variant of this, each using some of mixins like filters
// some posts notifiers might not be filterable for example but be refreshable etc.
final mainPostsProvider =
    StateNotifierProvider((ref) => PostsNotifier(PostsState.initial));

// you can easily override this with mainPostsProvider or anyother notifier mixing postsfilternotifier
// this will give access to only filters state and changefilters function
final filtersProvider = StateNotifierProvider<PostsFilterNotifier, PostsFilter>(
  (ref) => throw UnimplementedError(),
);
```
Using the new Notifier class, however, there's no unified class to depend on. It's not as easy to write shared logic and use providerscopes. Provider's lifecycle is directly tied to notifier implementation.
I can be wrong (and probably am) and also StateNotifier is still there and usable, but it would be nice to be able to write notifiers regardless of the lifecycle.

**Reply by abigotado** [reply]
@raminsafaei As it was mentioned previously by Remi you can use mixins but a little bit differently

Something like that:
```dart
mixin PostsFilter {
  bool get includeRead;
}

mixin PostsFilterMixin<S extends PostsFilter> {
  /// State from the class that uses the [PostsFilterMixin].
  State<AsyncValue<T>> get state;
  set state(State<AsyncValue<T> value);

  /// Provider reference from the class that uses the [PostsFilterMixin].
 Ref<AsyncValue<S>> get ref;

  void changeFilters();
}

class PostsState with PostsFilter {
  PostsState(this.posts, this.includeRead);

  final List<String> posts;
  @override
  final bool includeRead;

  static PostsState initial = PostsState([], true);
}

@riverpod
class PostsNotifier extends _$PostsNotifier with PostsFilterMixin<PostsState> {
  @override
  FutureOr<PostsState> build() {
    return loadPosts();
  }

  @override
  void changeFilters() {}
}
```

I have implemented this way of creating mixins for notifiers in my projects after Remi's post. And everything works well.

**Reply by ghost** [reply]
@abigotado thanks for response and code sample. I've read Remi's response before and have implemented a very similar thing.
My main problem was polymorphism and overriding notifiers. As I've mentioned before, using StateNotifiers It's easy to use provider scopes and give parts of a widget tree access to only parts of a bigger notifier. We could have a modal to change posts filters without giving it the whole notifier with dozens of methods and state, you could simply give it PostsFilterNotifier with PostsFilter as its state (as a StateNotifier), and it would just work. With this mixins and state etc. it's technically possible to achieve similar thing, but it's not a Notifier.
Of course in this example it's not that big of a deal, but you can imagine how handy it can be in more complex situations.
Also, I believe it's discouraged to read state outside of read/watch of notifier's state (it's protected). Can it lead to unintended rebuilds? (if I use PostsFilterMixin as you've written above)

**Comment by Dean-Spotec** [comment]
to encapsulate paging loading widget，[riverpod_paging_utils](https://pub.dev/packages/riverpod_paging_utils) uses the mixin solution. [riverpod_infinite_scroll](https://pub.dev/packages/riverpod_infinite_scroll) use the StateNotifier solution
